### PR TITLE
[Feature] Datadog profiler

### PR DESCRIPTION
**Pull request by ZeljkoBenovic**
This PR adds Datadog profiler into the codebase, so that it can be easily enabled on demand.
Profiler enables users to see profiles and tracing stats on their Datadog portal, but it requires some configuration of Datadog agent on the node. 

APM ( tracing ) should be enabled in Datadog agent configuration yaml file:
```
apm_config:

  enabled: true
  env: polygon-edge # this sets the environment name that will be shown on DD portal
```

To enable the profiler, a user must simply set `DD_PROFILING_ENABLED="True"` environment variable.
To set the IP and port that the DataDog agent, use `DD_AGENT_HOST` and `DD_TRACE_AGENT_PORT` environment variables.    
By default, it will try communicate with the agent on `localhost:8126`.

These env vars are now consistent with the DD doc https://docs.datadoghq.com/tracing/trace_collection/dd_libraries/go/?tab=containers#configure-the-datadog-agent-for-apm

Additional parameters can be set by using env vars stated in the official Datadog doc https://docs.datadoghq.com/profiler/enabling/go/

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have updated the official documentation
- [ ] I have added sufficient documentation in code

## Testing

- [ ] I have tested this code with the official test suite
- [x] I have tested this code manually

### Manual tests

Manual tests would require that the tester has access to Datadog portal.
Just set the `DD_ENABLE` env var, configure Datadog agent on the same node, and check Datadog portal if you can see your profile.

# Documentation update

Public doc will soon with the explanation on how to set this up will follow

# Additional comments

Fixes EDGE-764

### CODE REVIEW

File: server/server_metrics.go [hunk 1]
```diff
@@ -29,3 +33,45 @@ func metricProvider(nameSpace string, chainID string, metricsRequired bool) *ser
...
+	}
+
+	// start the tracer
+	tracer.Start()
+	s.logger.Info("DataDog profiler started")
+
+	return nil
+}
+
+func (s *Server) CloseDataDogProfiler() {
```
Comment (zivkovicmilos) [inline server/server_metrics.go]: This shouldn't be an exported method, no need for that

**Review by zivkovicmilos — APPROVED** [review]
Looks great 💯

File: server/server_metrics.go (hunk 1 above)
Comment (ZeljkoBenovic) [reply server/server_metrics.go]: unexported in be46b7199f54ae1cfd825dc31d9500004f810875

**Review by 0xAleksaOpacic — APPROVED** [review]
Code looks good to me, WHy are we having issues with CLAssistant first time seeing this

**Comment by 0xAleksaOpacic** [comment]
Can you add a comment: I have read the CLA Document and I hereby sign the CLA

File: server/server_metrics.go [hunk 2]
```diff
@@ -29,3 +33,45 @@ func metricProvider(nameSpace string, chainID string, metricsRequired bool) *ser
...
+func (s *Server) closeDataDogProfiler() {
+	defer func() {
+		s.logger.Debug("closing DataDog profiler")
+		profiler.Stop()
+	}()
+
+	defer func() {
+		s.logger.Debug("closing DataDog tracer")
+		tracer.Stop()
+	}()
```
Comment (Kourin1996) [inline server/server_metrics.go]: Is there any reason to use defer here?

File: server/server_metrics.go [hunk 3]
```diff
@@ -29,3 +33,45 @@ func metricProvider(nameSpace string, chainID string, metricsRequired bool) *ser
 		txpool:    txpool.NilMetrics(),
 	}
 }
+
+// enableDataDogProfiler enables DataDog profiler. Enable it by setting DD_ENABLE env var.
+// Additional parameters can be set with env vars (DD_) - https://docs.datadoghq.com/profiler/enabling/go/
+func (s *Server) enableDataDogProfiler() error {
+	if os.Getenv("DD_ENABLE") == "" {
```
Comment (Kourin1996) [inline server/server_metrics.go]: Why not check the value?

**Review by Kourin1996 — APPROVED** [review]

File: server/server_metrics.go (hunk 3 above)
Comment (epikichi) [reply server/server_metrics.go]: In addition, we can use the variable `DD_PROFILING_ENABLED` for consistency and anticipation for a similar option in the Datadog Java Profiler. https://docs.datadoghq.com/profiler/enabling/java/?tab=environmentvariables

File: server/server_metrics.go [hunk 4]
```diff
@@ -29,3 +33,45 @@ func metricProvider(nameSpace string, chainID string, metricsRequired bool) *ser
...
+		s.logger.Debug("DataDog profiler disabled, set DD_ENABLE env var to enable it.")
+
+		return nil
+	}
+
+	if err := profiler.Start(
+		// enable all profiles
```
Comment (epikichi) [inline server/server_metrics.go:61]: - We can look to add `profiler.WithAgentAddr()`  for containerized solutions, as the default agent address in the profiler is implicitly `localhost:8126`.

Reference: [1.30.0](https://github.com/DataDog/dd-trace-go/releases/tag/v1.30.0)

This behavior does not hold true for the tracer, as it takes up the `DD_AGENT_HOST` env var. The release has further explanation on why they introduced the behavior.

**Review by epikichi — CHANGES_REQUESTED** [review]

File: server/server_metrics.go (hunk 2 above)
Comment (ZeljkoBenovic) [reply server/server_metrics.go]: You're right. Defers here makes no sense. Removed e0a262a286b1bb9df18d4811b4c28f8f359f0bd0

File: server/server_metrics.go (hunk 4 above)
Comment (ZeljkoBenovic) [reply server/server_metrics.go:61]: Added the option to set IP and port, using env vars. `DD_PROF_IP` and `DD_PROF_PORT` in  e0a262a286b1bb9df18d4811b4c28f8f359f0bd0

Comment (ZeljkoBenovic) [reply server/server_metrics.go:61]: Changed env var to enable DD agent to `DD_PROFILING_ENABLED` and to define the agent host `DD_AGENT_HOST` and port `DD_TRACE_AGENT_PORT`

**Review by vcastellm — COMMENTED** [review]
The approach looks good in general, but I don't feel confortable with adding features using only env vars, I mean, ideally for me, this can be configurable in CLI params, config file and env vars.

**Comment by epikichi** [comment]
> The approach looks good in general, but I don't feel confortable with adding features using only env vars, I mean, ideally for me, this can be configurable in CLI params, config file and env vars.

I agree as well. Let's include the CLI and config equivalent to this PR. @ZeljkoBenovic 

Ideally, we can offload the implementation to https://github.com/spf13/viper @zivkovicmilos :eyes:

**Comment by ZeljkoBenovic** [comment]
@vcastellm The primary idea was not to clutter the UX with additional flags as DataDog is not a feature that is so commonly used as opposed to Prometheus for example.    
Secondly, we wanted to be aligned with their docs on how to setup things, for example https://docs.datadoghq.com/tracing/trace_collection/dd_libraries/go/?tab=containers#configure-the-datadog-agent-for-apm where they are using env vars.     
Additionally, this is not a standalone feature, it still requires DataDog Agent installed and configured on the host ( or as a remote instance ) 

If you guys think that we need flags/config file parameters for this, it is very easy to add them.

File: server/server_metrics.go (hunk 4 above)
Comment (epikichi) [reply server/server_metrics.go:61]: Looks good now. Thanks!

**Review by epikichi — APPROVED** [review]

**Review by vcastellm — APPROVED** [review]
LGTM

**Pull request merged**